connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java

```diff
@@ -1592,6 +1711,16 @@ public void onSessionKeyUpdate(SessionKey sessionKey) {
                 }
             }
         }
+
+        @Override
+        public void onRestartRequest(RestartRequest request) {
+            log.info("Received and enqueuing {}", request);
+
+            synchronized (DistributedHerder.this) {
+                pendingRestartRequests.put(request.connectorName(), request);
```

Comment:
One thing about using a map is that we're always saving the last restart request, rather than the restart request that will (potentially) have the greatest impact. 

Consider two restart requests are submitted at about the same time:
1. RestartRequest 1 requests that everything is restarted (onlyFailed=false) for connector Foo
2. RestartRequest 2 requests that only failed instances are restarted (onlyFailed=true) for connector Foo

If this method saw RestartRequest 2 before RestartRequest 1, then everything would be restarted as expected. But if this method saw RestartRequest 1 first, then only the failed tasks would be restarted.

Seems like we should keep the most impactful one: RestartRequest 1 in this example. WDYT?

Reply:
@rhauch  I agree and implemented it, Could you please review the below commit and see if it suffices? https://github.com/apache/kafka/pull/10822/commits/cb5afec3af37d60abae988c58e8180440ac97a58